File: packages/node/src/receiver.js
/*
 * receiver.js — the Express side: where the broker POSTs.
 *
 * bjmsg pushes. A subscription names a callback URL, the broker POSTs
 * each batch to it, and the HTTP response is the acknowledgement — so a
 * subscriber is a server, and this is that server.
 *
 * One Express app serves every subscription this client has, each on its
 * own path under `mountPath`. Routes are not added and removed as
 * subscriptions come and go (Express has no good way to remove one);
 * instead a single parameterised route dispatches from a table, which
 * also makes an unknown consumer a clean 404 rather than a stack trace.
 */
import http from 'node:http';

import express from 'express';

import { MEDIA_TYPE } from './transport.js';
import { parseDelivery, deliveryInfo } from './protocol.js';

const DEFAULT_BODY_LIMIT = '8mb';

export class Receiver {
  #app;
  #server;
  #own;                 /* we created the app, so we own its lifecycle */
  #routes = new Map();  /* consumer -> { token, handle } */

  /**
   * @param {object}  [opts]
   * @param {number}  [opts.port]       0 picks a free one
   * @param {string}  [opts.host]       what to bind; defaults to the
   *                                    address this host reaches the
   *                                    broker from
   * @param {string}  [opts.mountPath]  path prefix for delivery routes
   * @param {import('express').Express} [opts.app]
   *        an existing app to mount into, instead of creating one. You
   *        keep responsibility for listening and closing it.
   * @param {string|number} [opts.bodyLimit]
   */
  constructor({
    port = 0,
    host,
    mountPath = '/bjmsg',
    app = null,
    bodyLimit = DEFAULT_BODY_LIMIT,
  } = {}) {
    this.port = port;
    this.host = host;
    this.mountPath = mountPath.replace(/\/$/, '');
    this.#own = app === null;
    this.#app = app ?? express();

    /*
     * Raw bytes, not JSON: the body is binjson and Express must not try
     * to interpret it. `type` is the media type the broker sends, so any
     * other body parser the host app has installed is left alone.
     */
    this.#app.post(
      `${this.mountPath}/:consumer`,
      express.raw({ type: MEDIA_TYPE, limit: bodyLimit }),
      (req, res) => this.#deliver(req, res),
    );
  }

  /** The Express app, for mounting your own routes alongside. */
  get app() {
    return this.#app;
  }

  get listening() {
    return Boolean(this.#server?.listening);
  }

  /** Start listening. No-op when an existing app was supplied. */
  async listen(host = this.host) {
    if (!this.#own || this.#server) return this;
    this.#server = http.createServer(this.#app);
    /*
     * The broker holds this connection open and uses it whenever there
     * is something to send, which on a quiet subject may be a long time.
     * A short idle timeout would close a connection that is working
     * exactly as intended.
     */
    this.#server.keepAliveTimeout = 5 * 60 * 1000;
    this.#server.headersTimeout = 5 * 60 * 1000 + 1000;

    await new Promise((resolve, reject) => {
      this.#server.once('error', reject);
      this.#server.listen(this.port, host ?? '0.0.0.0', () => {
        this.#server.removeListener('error', reject);
        this.port = this.#server.address().port;
        resolve();
      });
    });
    return this;
  }

  async close() {
    this.#routes.clear();
    if (!this.#own || !this.#server) return;
    const server = this.#server;
    this.#server = null;
    await new Promise((resolve) => {
      server.close(resolve);
      /*
       * The broker holds its delivery connection open — that is the
       * design — and server.close() waits for every open connection to
       * end of its own accord. Without this it waits forever.
       */
      server.closeAllConnections?.();
    });
  }

  /** Where the broker should POST for `consumer`. */
  callbackUrl(consumer, advertisedHost) {
    const host = advertisedHost.includes(':') && !advertisedHost.startsWith('[')
      ? `[${advertisedHost}]`   /* an IPv6 literal needs brackets to carry a port */
      : advertisedHost;
    return `http://${host}:${this.port}${this.mountPath}/${consumer}`;
  }

  register(consumer, token, handle) {
    this.#routes.set(consumer, { token, handle });
  }

  unregister(consumer) {
    this.#routes.delete(consumer);
  }

  get size() {
    return this.#routes.size;
  }

  async #deliver(req, res) {
    const route = this.#routes.get(req.params.consumer);
    if (!route) {
      res.status(404).type('text/plain').send('no such subscription here\n');
      return;
    }

    /*
     * The token proves the POST came from the broker we registered with.
     * It is the only thing that does: anything that can reach this port
     * can connect to it, and a subscriber that took whatever arrived
     * would accept messages from anywhere.
     */
    if (route.token && req.get('authorization') !== `Bearer ${route.token}`) {
      res.status(401).type('text/plain').send('bad or missing bearer token\n');
      return;
    }

    if (!Buffer.isBuffer(req.body) || req.body.length === 0) {
      res.status(400).type('text/plain').send('empty delivery\n');
      return;
    }

    let messages;
    const info = deliveryInfo(req.headers);
    try {
      messages = parseDelivery(req.body);
    } catch {
      res.status(400).type('text/plain').send('malformed delivery\n');
      return;
    }

    await route.handle(messages, info, res);
  }
}

File: packages/python/bjmsg/receiver.py
"""The Flask side: where the broker POSTs.

bjmsg pushes. A subscription names a callback URL, the broker POSTs each
batch to it, and the HTTP response is the acknowledgement — so a
subscriber is a server, and this is that server.

One Flask app serves every subscription a client has, each on its own
path under `mount_path`. Routes are not added and removed as
subscriptions come and go; a single parameterised route dispatches from a
table, which also makes an unknown consumer a clean 404.
"""
from __future__ import annotations

import threading
from socketserver import ThreadingMixIn
from typing import Callable, Dict, Optional, Tuple
from wsgiref.simple_server import WSGIRequestHandler, WSGIServer, make_server

from flask import Flask, Response, request

from .protocol import delivery_info

DEFAULT_BODY_LIMIT = 8 * 1024 * 1024


class _Handler(WSGIRequestHandler):
    """
    HTTP/1.1, so the connection survives a response.

    wsgiref speaks HTTP/1.0 by default and closes after every exchange,
    which would make the broker reconnect for each delivery — losing the
    one kept-alive connection per subscriber that the whole design is
    arranged around. Flask sets Content-Length on every response, which
    is what HTTP/1.1 needs to keep the framing unambiguous.
    """

    protocol_version = "HTTP/1.1"
    #: wsgiref writes the status line, the headers and the body in
    #: separate socket writes. With Nagle on, the kernel holds the later
    #: ones waiting for an ACK — which is milliseconds added to every
    #: acknowledgement, on a path whose whole point is not waiting.
    disable_nagle_algorithm = True

    def log_message(self, fmt, *args):
        """A delivery is not news; the library logs what matters."""


class _ThreadingWSGIServer(ThreadingMixIn, WSGIServer):
    """One thread per connection, so one subscription's handler cannot
    hold up another's delivery. Within a subscription the broker sends
    one batch at a time regardless, so message order is not at risk."""

    daemon_threads = True
    #: The broker parks a connection here between deliveries, which on a
    #: quiet subject may be a long time; nothing should reap it.
    allow_reuse_address = True


class Receiver:
    """
    The HTTP server deliveries arrive on.

    Pass `app` to mount into a Flask app you already have — the routes go
    on, and listening stays yours. Otherwise this starts its own server on
    a background thread.
    """

    def __init__(
        self,
        *,
        port: int = 0,
        host: Optional[str] = None,
        mount_path: str = "/bjmsg",
        app: Optional[Flask] = None,
        body_limit: int = DEFAULT_BODY_LIMIT,
    ):
        self.port = port
        self.host = host
        self.mount_path = mount_path.rstrip("/")
        self.own = app is None
        self.app = app or Flask("bjmsg.receiver")
        self.app.config.setdefault("MAX_CONTENT_LENGTH", body_limit)

        self._routes: Dict[str, Tuple[str, Callable]] = {}
        self._lock = threading.Lock()
        self._server = None
        self._thread: Optional[threading.Thread] = None

        self.app.add_url_rule(
            f"{self.mount_path}/<consumer>",
            endpoint=f"bjmsg_deliver_{id(self)}",
            view_func=self._deliver,
            methods=["POST"],
        )

    # -- lifecycle ---------------------------------------------------------

    @property
    def listening(self) -> bool:
        return self._server is not None

    def listen(self, host: Optional[str] = None) -> "Receiver":
        """Start listening. A no-op when an existing app was supplied."""
        if not self.own or self._server is not None:
            return self
        bind = host or self.host or "0.0.0.0"
        self._server = make_server(
            bind, self.port, self.app,
            server_class=_ThreadingWSGIServer, handler_class=_Handler,
        )
        self.port = self._server.server_port
        self._thread = threading.Thread(
            target=self._server.serve_forever,
            name="bjmsg-receiver",
            daemon=True,
        )
        self._thread.start()
        return self

    def close(self) -> None:
        with self._lock:
            self._routes.clear()
        if not self.own or self._server is None:
            return
        server, thread = self._server, self._thread
        self._server = self._thread = None
        server.shutdown()
        server.server_close()
        if thread is not None:
            thread.join(timeout=5)

    # -- subscriptions ------------------------------------------------------

    def callback_url(self, consumer: str, advertised_host: str) -> str:
        host = advertised_host
        if ":" in host and not host.startswith("["):
            host = f"[{host}]"          # an IPv6 literal needs brackets
        return f"http://{host}:{self.port}{self.mount_path}/{consumer}"

    def register(self, consumer: str, token: str, handle: Callable) -> None:
        with self._lock:
            self._routes[consumer] = (token, handle)

    def unregister(self, consumer: str) -> None:
        with self._lock:
            self._routes.pop(consumer, None)

    def __len__(self) -> int:
        return len(self._routes)

    # -- the route ----------------------------------------------------------

    def _deliver(self, consumer: str):
        with self._lock:
            route = self._routes.get(consumer)
        if route is None:
            return Response("no such subscription here\n", status=404,
                            mimetype="text/plain")

        token, handle = route
        # The token proves the POST came from the broker we registered
        # with. It is the only thing that does: anything that can reach
        # this port can connect to it, and a subscriber that took
        # whatever arrived would accept messages from anywhere.
        if token and request.headers.get("Authorization") != f"Bearer {token}":
            return Response("bad or missing bearer token\n", status=401,
                            mimetype="text/plain")

        body = request.get_data()
        if not body:
            return Response("empty delivery\n", status=400, mimetype="text/plain")

        info = delivery_info(dict(request.headers))
        return handle(body, info)
